matchers/parse/lexer_test.go

```diff
@@ -0,0 +1,564 @@
+		input: "=",
+		expected: []Token{{
+			Kind:  TokenOperator,
+			Value: "=",
+			Position: Position{
+				OffsetStart: 0,
+				OffsetEnd:   1,
+				ColumnStart: 0,
+				ColumnEnd:   1,
+			},
+		}},
+	}, {
+		name:  "not equals operator",
+		input: "!=",
+		expected: []Token{{
+			Kind:  TokenOperator,
+			Value: "!=",
+			Position: Position{
+				OffsetStart: 0,
+				OffsetEnd:   2,
+				ColumnStart: 0,
+				ColumnEnd:   2,
+			},
+		}},
+	}, {
+		name:  "matches regex operator",
+		input: "=~",
+		expected: []Token{{
+			Kind:  TokenOperator,
+			Value: "=~",
+			Position: Position{
+				OffsetStart: 0,
+				OffsetEnd:   2,
+				ColumnStart: 0,
+				ColumnEnd:   2,
+			},
+		}},
+	}, {
+		name:  "not matches regex operator",
+		input: "!~",
+		expected: []Token{{
+			Kind:  TokenOperator,
+			Value: "!~",
+			Position: Position{
+				OffsetStart: 0,
+				OffsetEnd:   2,
+				ColumnStart: 0,
+				ColumnEnd:   2,
+			},
+		}},
+	}, {
+		name:  "unexpected $",
+		input: "$",
+		err:   "0:1: $: invalid input",
+	}, {
+		name:  "unexpected emoji",
+		input: "🙂",
+		err:   "0:1: 🙂: invalid input",
+	}, {
+		name:  "unexpected unicode letter",
+		input: "Σ",
+		err:   "0:1: Σ: invalid input",
+	}, {
+		name:  "unexpected : at start of ident",
+		input: ":hello",
+		err:   "0:1: :: invalid input",
+	}, {
+		name:  "unexpected $ in ident",
+		input: "hello$",
+		expected: []Token{{
+			Kind:  TokenIdent,
+			Value: "hello",
+			Position: Position{
+				OffsetStart: 0,
+				OffsetEnd:   5,
+				ColumnStart: 0,
+				ColumnEnd:   5,
+			},
+		}},
+		err: "5:6: $: invalid input",
+	}, {
+		name:  "unexpected unicode letter in ident",
+		input: "helloΣ",
+		expected: []Token{{
+			Kind:  TokenIdent,
+			Value: "hello",
+			Position: Position{
+				OffsetStart: 0,
+				OffsetEnd:   5,
+				ColumnStart: 0,
+				ColumnEnd:   5,
+			},
+		}},
+		err: "5:6: Σ: invalid input",
+	}, {
+		name:  "unexpected emoji in ident",
+		input: "hello🙂",
+		expected: []Token{{
+			Kind:  TokenIdent,
+			Value: "hello",
+			Position: Position{
+				OffsetStart: 0,
+				OffsetEnd:   5,
+				ColumnStart: 0,
+				ColumnEnd:   5,
+			},
+		}},
+		err: "5:6: 🙂: invalid input",
+	}, {
+		name:  "invalid operator",
+		input: "!",
+		err:   "0:1: unexpected end of input, expected one of '=~'",
+	}, {
+		name:  "another invalid operator",
+		input: "~",
+		err:   "0:1: ~: invalid input",
+	}, {
+		name:  "unexpected $ in operator",
+		input: "=$",
+		expected: []Token{{
+			Kind:  TokenOperator,
+			Value: "=",
+			Position: Position{
+				OffsetStart: 0,
+				OffsetEnd:   1,
+				ColumnStart: 0,
+				ColumnEnd:   1,
+			},
+		}},
+		err: "1:2: $: invalid input",
+	}, {
+		name:  "unexpected ! after operator",
+		input: "=!",
+		expected: []Token{{
+			Kind:  TokenOperator,
+			Value: "=",
+			Position: Position{
+				OffsetStart: 0,
+				OffsetEnd:   1,
+				ColumnStart: 0,
+				ColumnEnd:   1,
+			},
+		}},
+		err: "1:2: unexpected end of input, expected one of '=~'",
```

Comment:
It's because the lexer has found a valid operator `=`, and then the start of another operator `!`. However, `!` is not valid unless its followed with either a `=` for `!=` or a `~` for `!~`, so it expects either a `=` or a `~` after the `!`.

Here is an example of how it's lexed using [`matchers-cli --just-lex`](https://github.com/grobinson-grafana/matchers-cli):

```
=!
(Op) '='
1:2: unexpected end of input, expected one of '=~'
```

Reply:
Makes sense, thanks for the explanation. Should we fix the message? It reads "one of" when there's only one option.

Reply:
No no there are two options! It expects either a `=` or a `~` after the `!` but found neither. I chose not to comma separate them because when a comma is expected it doesn't make sense:

Comma separated, no comma expected:

```
1:2: unexpected end of input, expected one of '=,~'
```

Comma separated, one of `=`, `,` or `~` expected:

```
1:2: unexpected end of input, expected one of '=,,,~'
```

Reply:
Ah, now I get it! Yeah, I think we can improve of this.

What about?

```
1:2: unexpected end of input, perhaps you meant '==' or '=~'?
```

Reply:
Yes, but it will instead be:

```
1:2: unexpected end of input, expect one of '=' or '~'.
```

as the lexer has consumed a valid character `!` and is now looking for either an `=` or `~` to complete it.